src/chainstate/stacks/index/marf.rs

```diff
@@ -0,0 +1,2228 @@
+/*
+ copyright: (c) 2013-2019 by Blockstack PBC, a public benefit corporation.
+
+ This file is part of Blockstack.
+
+ Blockstack is free software. You may redistribute or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation, either version 3 of the License or
+ (at your option) any later version.
+
+ Blockstack is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY, including without the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with Blockstack. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use std::fmt;
+use std::error;
+use std::io;
+use std::io::{
+    Read,
+    Write,
+    Seek,
+    SeekFrom,
+    Cursor
+};
+
+use std::marker::PhantomData;
+use std::fs;
+
+use sha2::Digest;
+
+use chainstate::burn::BlockHeaderHash;
+
+use chainstate::stacks::index::bits::{
+    get_node_hash,
+    get_node_hash_bytes
+};
+
+use chainstate::stacks::index::node::{
+    TrieNodeID,
+    TrieNodeType,
+    TrieNode,
+    TrieNode4,
+    TrieNode16,
+    TrieNode48,
+    TrieNode256,
+    TrieLeaf,
+    TriePtr,
+    TRIEPTR_SIZE,
+    TrieCursor,
+    TriePath,
+    is_backptr,
+    set_backptr,
+    clear_backptr,
+};
+
+use chainstate::stacks::index::storage::{
+    read_all,
+    write_all,
+    fseek,
+    fseek_end,
+    TrieStorage,
+    TrieFileStorage
+};
+
+use chainstate::stacks::index::{
+    TrieHash,
+    TRIEHASH_ENCODED_SIZE,
+    fast_extend_from_slice,
+    MARFValue
+};
+
+use chainstate::stacks::index::trie::{
+    Trie,
+};
+
+use chainstate::stacks::index::Error as Error;
+
+use util::log;
+
+
+/// Merklized Adaptive-Radix Forest -- a collection of Merklized Adaptive-Radix Tries.
+pub struct MARF<S>
+where
+    S: TrieStorage + Seek
+{
+    storage: S,
+    open_chain_tip: Option<BlockHeaderHash>
+}
+
+impl<S> MARF<S>
+where
+    S: TrieStorage + Seek
+{
+
+    // helper method for walking a node's backpr
+    fn walk_backptr(s: &mut S, start_node: &TrieNodeType, chr: u8, c: &mut TrieCursor) -> Result<(TrieNodeType, TrieHash, TriePtr, u32), Error> {
```

Comment:
Not exactly a bug, but I think it affects readability of this code: there are a lot of single letter and three letter function arguments in this file (many of which are very close to other arguments (`c` vs. `chr`, `s` vs. `start_node`)).

Reply:
I can take a crack at cleaning this up.

Reply:
This is addressed now.  In particular, `s: S` is now `storage: S`, and `c: &TrieCursor` (or `c: &mut TrieCursor`) is now `cursor: ...`.